tracing: add username to the session table

wmitros opened the pull request:
In order to improve observability, add a userspace field to the the system_traces.sessions table.

Fixes #6737.

haaawk commented:
"In order to improve observability, add a userspace field to the the system_traces.sessions table." Shouldn't it be "username field"?

haaawk commented on tracing/trace_keyspace_helper.cc:
```diff
@@ -87,6 +87,7 @@ trace_keyspace_helper::trace_keyspace_helper(tracing& tr)
                                   "started_at timestamp,"
                                   "request_size int,"
                                   "response_size int,"
+                                  "username text,"
```
Will this work after upgrading an existing cluster? There's a "IF NOT EXISTS" in line 79 so I would expect it won't. Please run a version of Scylla without your change and use tracing on it. Then start a version of Scylla with your change and verify everything works as expected and the new field is there and filled correctly.

wmitros replied:
It doesn't work with upgrading as it is right now. I don't see how we can remember the usernames for requests before upgrading, should this field be filled with nulls for them?

haaawk replied:
Filling the new column up with nulls for old entries is alright, I think.

avikivity replied:
You'll have a race between preparing the statement and altering the table to add username. So I suggest to try to prepare with username, and fall back to without username. There's a chance that executing a query with an extra parameter won't hurt.

haaawk commented on table_helper.cc:
```diff
@@ -31,7 +32,22 @@ future<> table_helper::setup_table() const {
+                            static_pointer_cast<cql3::statements::alter_table_statement>(
+                                            parsed_cf_stmt->prepare(db, qp.get_cql_stats())->statement);
+
+            // Instead of checking if the alteration is possible, we try it, and in case it fails we change nothing
```
Why is it ok to just fail and not retry? What are the reasons the alternation could be impossible? Do we have a guarantee that eventually the cluster will move to the new schema that contains username?

wmitros replied:
The reason for the fail is that the alteration has already been done, which could have happened in the same or a different node and I can't actually see any other reason - in the previous version we assumed that the schema is up-to-date here, so in this version it should be in the old state or in a state after altering the old state. (We could manually destroy the table but I don't think we're implementing repairing of such damage)
The cluster should move to the new schema after the first call(or concurrent calls) of `statement->announce_migration`, which is called (indirectly) as early as during the init phase.
Note that this assumes that we're upgrading by adding new columns - current version wouldn't work well with dropping or renaming.
Please correct me if I missed anything.

wmitros replied:
After researching a bit more it looks like all nodes will get the final schema after the last node that didn't get the schema from gossip sends its version with latest timestamp to all nodes.(It probably shouldn't happen in practice - every node aside from the first should get the scheme on joining the cluster)

haaawk commented:
I would swap the order of the commits. Or maybe extract the change to table_helper to separate commit and merge the rest into the second commit.

slivne commented:
we have an external script as well to alter tracing tables - its old and is an issue in itself

https://github.com/scylladb/scylla/blob/master/fix_system_distributed_tables.py

at the time the thought was to ask users to run it after upgrade.

haaawk commented:
> we have an external script as well to alter tracing tables - its old and is an issue in itself
> 
> https://github.com/scylladb/scylla/blob/master/fix_system_distributed_tables.py
> 
> at the time the thought was to ask users to run it after upgrade.

It keeps the codebase cleaner but is probably more of a hassle for users. Do we want to stick to the script approach or is it ok to make it automatic as Wojciech has done in the latest revision?

slivne commented:
I would guess avi prefers the automatic approach - but better make sure -
and while you are at it - lets also file an issue about this and fix this
as well.

On Tue, Jul 14, 2020 at 6:36 PM Piotr Jastrzębski <notifications@github.com>
wrote:

> we have an external script as well to alter tracing tables - its old and
> is an issue in itself
>
>
> https://github.com/scylladb/scylla/blob/master/fix_system_distributed_tables.py
>
> at the time the thought was to ask users to run it after upgrade.
>
> It keeps the codebase cleaner but is probably more of a hassle for users.
> Do we want to stick to the script approach or is it ok to make it automatic
> as Wojciech has done in the latest revision?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/6827#issuecomment-658251014>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2OCCDNEENBDWD4UUXWB2LR3R3PPANCNFSM4OYRIMCQ>
> .
>

haaawk commented:
> I would guess avi prefers the automatic approach - but better make sure - and while you are at it - lets also file an issue about this and fix this as well.


What do you mean by *this*?

slivne commented:
If avi wants an automatic alter

lets remove the script I shared and do its update also in this manner.

On Tue, Jul 14, 2020 at 6:39 PM Piotr Jastrzębski <notifications@github.com>
wrote:

> I would guess avi prefers the automatic approach - but better make sure -
> and while you are at it - lets also file an issue about this and fix this
> as well.
>
> What do you mean by *this*?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/6827#issuecomment-658252645>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2OCCANVYKP6AVAECXTDZTR3R32XANCNFSM4OYRIMCQ>
> .
>

haaawk commented:
> If avi wants an automatic alter lets remove the script I shared and do its update also in this manner.

Ok. I will create an issue and assign @wmitros to it so that he could hit both birds with a single stone.

slivne commented:
Again, lets get Avi blessing on the automatic method first.

On Tue, Jul 14, 2020 at 6:45 PM Piotr Jastrzębski <notifications@github.com>
wrote:

> If avi wants an automatic alter lets remove the script I shared and do its
> update also in this manner.
>
> Ok. I will create an issue and assign @wmitros
> <https://github.com/wmitros> to it so that he could hit both birds with a
> single stone.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylla/pull/6827#issuecomment-658255793>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA2OCCEK2RTAUBONE3QHTNLR3R4QHANCNFSM4OYRIMCQ>
> .
>

slivne commented:
I think we may need to consider doing the changes of schemas only after the
whole cluster is upgraded or else we will not really be able to rollback.

On Tue, Jul 14, 2020 at 6:47 PM Shlomi Livne <shlomi@scylladb.com> wrote:

> Again, lets get Avi blessing on the automatic method first.
>
> On Tue, Jul 14, 2020 at 6:45 PM Piotr Jastrzębski <
> notifications@github.com> wrote:
>
>> If avi wants an automatic alter lets remove the script I shared and do
>> its update also in this manner.
>>
>> Ok. I will create an issue and assign @wmitros
>> <https://github.com/wmitros> to it so that he could hit both birds with
>> a single stone.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scylladb/scylla/pull/6827#issuecomment-658255793>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AA2OCCEK2RTAUBONE3QHTNLR3R4QHANCNFSM4OYRIMCQ>
>> .
>>
>

haaawk commented:
> > If avi wants an automatic alter lets remove the script I shared and do its update also in this manner.
> 
> Ok. I will create an issue and assign @wmitros to it so that he could hit both birds with a single stone.

On a second thought @slivne - do we really have to do anything? The script hasn't been changed for over a year. Maybe it makes sense to leave it for migrations from older clusters and just make sure we don't update it anymore? If I remember correctly, we allow only upgrades to the next version so making the script automatic would require backporting changes into multiple versions that were released during this year.

haaawk commented:
> > > If avi wants an automatic alter lets remove the script I shared and do its update also in this manner.
> > 
> > 
> > Ok. I will create an issue and assign @wmitros to it so that he could hit both birds with a single stone.
> 
> On a second thought @slivne - do we really have to do anything? The script hasn't been changed for over a year. Maybe it makes sense to leave it for migrations from older clusters and just make sure we don't update it anymore? If I remember correctly, we allow only upgrades to the next version so making the script automatic would require backporting changes into multiple versions that were released during this year.

I guess what I'm trying to say is that we can just drop the script with no changes needed to code in master.

haaawk commented:
> I think we may need to consider doing the changes of schemas only after the whole cluster is upgraded or else we will not really be able to rollback.
> […](#)
> On Tue, Jul 14, 2020 at 6:47 PM Shlomi Livne ***@***.***> wrote: Again, lets get Avi blessing on the automatic method first. On Tue, Jul 14, 2020 at 6:45 PM Piotr Jastrzębski < ***@***.***> wrote: > If avi wants an automatic alter lets remove the script I shared and do > its update also in this manner. > > Ok. I will create an issue and assign @wmitros > <https://github.com/wmitros> to it so that he could hit both birds with > a single stone. > > — > You are receiving this because you commented. > Reply to this email directly, view it on GitHub > <[#6827 (comment)](https://github.com/scylladb/scylla/pull/6827#issuecomment-658255793)>, > or unsubscribe > <https://github.com/notifications/unsubscribe-auth/AA2OCCEK2RTAUBONE3QHTNLR3R4QHANCNFSM4OYRIMCQ> > . >

It should be fine. The side effect could be an additional unused column in the table but it shouldn't prevent the cluster from working.

wmitros commented:
This version is actually still not working quite as intended and the commits will need rearranging, should I convert it to draft or somehow mark that it's not ready for review?

haaawk commented:
just prefix the title of the PR with RFC/Do not merge

vladzcloudius commented:
@wmitros I'm sorry for a late jumping onto this train but we've tried to do similar things in the past and decided not to do schema alterations from the tracing code due to unresolvable races.

If memory serves me well this was in the context of adding `scylla_parent_id` and `scylla_span_id` columns.

AFAIK nothing has drammaticaly changed in the way we handle schema changes since then and all those races are still where they were back then.

It was then decided to use an external script in order to adjust the system_traces schema: https://github.com/scylladb/scylla/blob/master/fix_system_distributed_tables.py

So, as far as I can see there are at least 2 ways you can proceed here:
1) Add a new column and adjust the script above.
2) Add the `username` value to `parameters` map.

Option 1 is probably better from the DBA perspective but Option 2 is a lot easier to implement and doesn't require schema adjustments.

So, whatever will make this get merged faster works for me.
@avikivity Please, comment.

avikivity commented:
I wanted an external script as to avoid adding more C++ code everwhere. We can consider a .cql file that will be executed by the cluster rather than going through the driver (and suffering from auth problems).

We still have the problems of races. 
 - which node executes the script (can be easily resolved once we have Raft)
 - use the new columns/tables vs. run the script (can be resolved by tolerating failures and falling back to pre-modification, but need to retry with the new schema eventually)

vladzcloudius commented:
> I wanted an external script as to avoid adding more C++ code everwhere. We can consider a .cql file that will be executed by the cluster rather than going through the driver (and suffering from auth problems).
> 
> We still have the problems of races.
> 
>     * which node executes the script (can be easily resolved once we have Raft)
> 
>     * use the new columns/tables vs. run the script (can be resolved by tolerating failures and falling back to pre-modification, but need to retry with the new schema eventually)

@avikivity The script is not executed automatically by the node today. 
Today a user has to go an execute it once after the upgrade that involved a schema change.
If needs this involves authentication.

Until we resolve races in the schema change path (using Paxos, Raft, etc.) we can't let this run automatically and having the user run it is an easy and a robust workaround.

wmitros commented:
Making the user run the script doesn't resolve the second race, right? We still need to use pre-modification statements until the script is executed because otherwise we wouldn't be able to collect tracing info during that period.

vladzcloudius commented:
> Making the user run the script doesn't resolve the second race, right? We still need to use pre-modification statements until the script is executed because otherwise we wouldn't be able to collect tracing info during that period.

Correct. That's exactly how it was and is so far and we are very well aware of this behaviour. 
This is not as bad as it sounds though because it clearly notifies the user that he/she needs to run the script in question (in case he/she forgot).
And once you ran it - you don't need to run it anymore (till the next schema change, which should be rare).

But all this could be a non-issue if you decide to add the username to the `parameters` map instead of adding them as a column.

wmitros commented:
I've added the username to the script and cql statements - the old statement is used if the new one fails. Checked manually if both statements work by executing statements with tracing enabled before and after executing the script and checking changes in the system_traces.sessions table(starting with outdated tables). Also ran scylla tests and cql tracing dtests.

haaawk commented on table_helper.cc:
```diff
@@ -95,6 +109,14 @@ future<> table_helper::insert(service::query_state& qs, noncopyable_function<cql
 }
 
+future<> table_helper::insert(service::query_state& qs, noncopyable_function<cql3::query_options (bool)> opt_maker) {
+    return cache_table_info(qs).then([this, &qs, opt_maker = std::move(opt_maker)] () mutable {
```
Why is mutable needed here?

wmitros replied:
Looks like I repeated the mistake from the other table_helper::insert overload

haaawk commented on table_helper.hh:
```diff
@@ -95,8 +97,27 @@ public:
+     */
+    template <typename OptMaker, typename... Args>
+    requires seastar::CanInvoke<OptMaker, Args...>
+    future<> insert(service::query_state& qs, OptMaker opt_maker, OptMaker opt_maker_fallback, Args... opt_maker_args) {
```
Where is this called? I couldn't find the user.

haaawk replied:
Ok. I see now.

haaawk commented on table_helper.cc:
```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
+        if (!_insert_cql_fallback) {
+            return make_exception_future(eptr);
+        }
+        if (_is_fallback_stmt && _prepared_stmt) {
```
Why is it ok to return under this condition?

wmitros replied:
The exception has been thrown if we couldn't prepare the non-fallback statement. We want to try to use the fallback statement now, so if we have it prepared, we use it.

haaawk replied:
I guess if we assume that none of the callbacks throws an exception then it should be ok.

haaawk replied:
Makes sense. Thanks for the explanation.

haaawk commented on table_helper.cc:
```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
+    if (!_prepared_stmt) {
         // if prepared statement has been invalidated - drop cached pointers
         _insert_stmt = nullptr;
+    } else if (!_is_fallback_stmt) {
```
I have a hard time parsing this if-else. I see following cases:
!_prepared_stmt -> we prepare
_prepared_stmt && !_is_fallback_stmt -> return because we have already prepared the statement
_prepared_stmt && _is_fallback_stmt -> we have fallback statement prepared. Why do we want to re-prepare in this case?

wmitros replied:
We've used the fallback statement because we couldn't use the regular statement (the table isn't updated yet). We want to start using the regular statement when the table gets updated, so we'll try to prepare it even if we have a successfully prepared fallback statement. If we fail again, we check for _prepared_stmt && _is_fallback_stmt to avoid re-preparing the fallback statement.

haaawk replied:
Thanks for the explanation. Probably worth leaving a comment like that in the code.

vladzcloudius replied:
And what if tomorrow we want to add one more column or N more? What will that be:
```
if (statement) ... else if (fallback1_statement)...else if (fallback2_statement)....if (fallbackN_statement)...
```

This obviously doesn't scale.
This fallback game leads to a very not-pretty spaghetti-like code hard to read and maintain.

If we believe that we want to add this new column let's simply add it and whoever haven't updated their schema will not be able to use Tracing until he/she does (updates the schema).

This is exactly how it works today and the only complain there is - there no enough documentation and some packaging didn't include the `fix_system_distributed_tables.py` script: https://github.com/scylladb/scylla/issues/4601

I don't see a reason to change an approach in general and this "fallback" proposal doesn't look like a good alternative to me either.

haaawk replied:
I'm ok with dropping the fallback and just relying on the script. @avikivity what about you? You proposed the fallback in https://github.com/scylladb/scylla/pull/6827#discussion_r454158997 so the question is whether you're ok with not doing it.

avikivity replied:
First, there will always be a window of time during cluster upgrade where we need the fallback.

Second, we don't enforce running the script, so we can be left with a cluster that doesn't support tracing.

haaawk replied:
> First, there will always be a window of time during cluster upgrade where we need the fallback.
We will need the fallback only if tracing is enabled or any table, right?
I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.

> 
> Second, we don't enforce running the script, so we can be left with a cluster that doesn't support tracing.

@vladzcloudius any counter arguments or should we just merge as it is? I like the simplest solution possible but if it's not sufficient maybe the complexity is worth it. How are the problems mentioned by @avikivity handled right now by the clients?

avikivity replied:
> > First, there will always be a window of time during cluster upgrade where we need the fallback.
> 
> We will need the fallback only if tracing is enabled on any table, right?

Yes.

> I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.

Application development and cluster operation can be completely separated. The cluster operator can even be a machine (the kubernetes operator). Even the application may decide to trace on its own (okay, that's not very likely). But it remains that forcing the application developer to synchronize with cluster operations is not good.

haaawk replied:
> > > First, there will always be a window of time during cluster upgrade where we need the fallback.
> > 
> > 
> > We will need the fallback only if tracing is enabled on any table, right?
> 
> Yes.
> 
> > I guess forcing users to ensure no queries with tracing are running during rolling upgrade is quite a burden and will back fire sooner or later.
> 
> Application development and cluster operation can be completely separated. The cluster operator can even be a machine (the kubernetes operator). Even the application may decide to trace on its own (okay, that's not very likely). But it remains that forcing the application developer to synchronize with cluster operations is not good.

It's hard to argue with this point. We should be doing all we can to automate the operations and not impose manual procedures and checklists. If I were to choose, I would go with a solution that does not involve user actions.

vladzcloudius replied:
@avikivity @haaawk The chance for such a scenario (devops are executing an upgrade while somebody debugs something at the same time using tracing is close to 0).
In a case of such an unlikely event during the window mentioned by Avi before the same engineer who debugs will probably have to wait till the end of the upgrade. - He'll probably have to wait for that due to multiple other reasons. Therefore there shouldn't be any drama if he forgets to stop traces and some of them fail.

But look from the other side - if we merge this fallback thingy we will have to maintain it.
What if we decide to add a few more columns in the future? How many fallback branches do we want to maintain?

I think it's safe enough not to worry about a few failed traces during the upgrade instead of merging this code which has a good potential to become a spaghetti.

haaawk replied:
Personally, I'm not in favor of the spaghetti code argument. If in the future we want/need to add more columns then we deal with it then. The code can always be refactored. The decision about having fallbacks can always be reverted unless we make it public (which we don't plan to do). Let's not over-engineer. At the moment it's a single `if` and that's what we're talking about. Hypothesizing is not a graceful base for arguments because it also powers argument like "what if in the future we have a potential client that wants to pay us 1B USD but requires tracing to always work". Also, we can remove the fallback code in the next release after the one that introduces the new column, no?

So far we have the following arguments:
For fallback:
1. Not having it opens up a can of worms related to human errors like DevOps team forgetting to run the script, cluster running for days, weeks, months, then having an issue which someone (potentially @vladzcloudius or someone less experienced) tries to debug with tracing but it fails. In such potentially stressful situation it is definitely something desirable. Not only will it slow down the process of solving the issue but could also hurt customer confidence in us.
2. It may be not ok for tracing to not work during the rolling upgrade.

Against fallback:
1. Code is simpler without callback
2. It may be ok for a query with tracing to fail (is it really?) for a short (is it really short) period of time the cluster is in rolling upgrade

I'm not listing the 'in the future' argument because I don't think it's a problem. The code can always be adjusted to future needs.
The existence of a proper solution based on transactions is kind of neutral because it means we will be able to remove the fallback code once we have transactions but no one really knows when that will be. In the meantime, we will have to deal with customer clusters for quite some time.

My suggestion is to merge the fallback code with a comment to remove it after one release (not the minor release like 4.1.1->4.1.2). That gives us most of the benefits. Tracing will work during a rolling upgrade and the code will still be simpler in the long run. The only thing we won't get is the protection in case of the situation the DevOps team forgets to run the script. In such case tracing will work until the next rolling upgrade (but not minor so until 4.1->4.2, not until 4.1.1->4.1.2) if they forget to run the script again. Not ideal but at least reduces the chances by giving the DevOps team few events (minor version upgrades and one non-minor upgrade) to think about running this script. @avikivity @vladzcloudius what do you think?

haaawk replied:
We could also check with the new version that the tracing table has the right schema and refuse to boot if not.

avikivity replied:
> @avikivity @haaawk The chance for such a scenario (devops are executing an upgrade while somebody debugs something at the same time using tracing is close to 0).
> In a case of such an unlikely event during the window mentioned by Avi above the engineer who debugs will probably have to wait till the end of the upgrade. - More than that, he'll probably have to wait due to multiple other reasons beyond tracing. Therefore there shouldn't be any drama if he forgets to stop traces and some of them fail.
> 
> But look from the other side - if we merge this fallback thingy we will have to maintain it.
> What if we decide to add a few more columns in the future? How many fallback branches do we want to maintain?
> 

As many as needed. We've added plenty of cluster features to make upgrades painless.

> I think it's safe enough not to worry about a few failed traces during the upgrade instead of merging this code which has a good potential to become spaghetti.

How did you arrive at "a few"? It could be a large number since cluster upgrade is not bounded in time, and there is nothing that will add the missing column automatically.

> 
> A proper way to address this is to make sure that a schema change is a transaction (on top of Paxos or Raft) and then trigger a schema change during the first boot after the upgrade.

This is not something we can do now.

avikivity replied:
> Personally, I'm not in favor of the spaghetti code argument. If in the future we want/need to add more columns then we deal with it then. The code can always be refactored. The decision about having fallbacks can always be reverted unless we make it public (which we don't plan to do). Let's not over-engineer. At the moment it's a single `if` and that's what we're talking about. Hypothesizing is not a graceful base for arguments because it also powers argument like "what if in the future we have a potential client that wants to pay us 1B USD but requires tracing to always work". Also, we can remove the fallback code in the next release after the one that introduces the new column, no?
> 
> So far we have the following arguments:
> For fallback:
> 
> 1. Not having it opens up a can of worms related to human errors like DevOps team forgetting to run the script, cluster running for days, weeks, months, then having an issue which someone (potentially @vladzcloudius or someone less experienced) tries to debug with tracing but it fails. In such potentially stressful situation it is definitely something not desirable. Not only will it slow down the process of solving the issue but could also hurt customer confidence in us.
> 2. It may be not ok for tracing to not work during the rolling upgrade.
> 
> Against fallback:
> 
> 1. Code is simpler without callback

IMO the extra cost in complexity is minor. An "if" statement is not something beyond our capability.

> 2. It may be ok for a query with tracing to fail (is it really?) for a short (is it really short?) period of time the cluster is in rolling upgrade

First, upgrade is not a short time.

Second, it is a stressful period, and having an extra source of failures is making it worse.

Third, we don't yet have an automatic way of adding that column.

> 
> I'm not listing the 'in the future' argument because I don't think it's a problem. The code can always be adjusted to future needs.
> The existence of a proper solution based on transactions is kind of neutral because it means we will be able to remove the fallback code once we have transactions but no one really knows when that will be. In the meantime, we will have to deal with customer clusters for quite some time.
> 
> **My suggestion is to merge the fallback code with a comment to remove it after one release** (not the minor release like 4.1.1->4.1.2). That gives us most of the benefits. Tracing will work during a rolling upgrade and the code will still be simpler in the long run. The only thing we won't get is the protection in case of the situation the DevOps team forgets to run the script. In such case tracing will work until the next rolling upgrade (but not minor so until 4.1->4.2, not until 4.1.1->4.1.2) if they forget to run the script again. Not ideal but at least reduces the chances by giving the DevOps team few events (minor version upgrades and one non-minor upgrade) to think about running this script. @avikivity @vladzcloudius what do you think?

That code should live on for a year, maybe two. Like all of the compatibility hacks. See 253a7640e3c770d939ae0024ee072a76d8d9c6cc which cleaned up two year old cluster features. This is no different.

I am confident in our ability to maintain that if statement, and perhaps even add another branch in the event that we add another column in the next year or two.

wmitros replied:
Considering there has been no discussion in the last couple of days, is it safe to assume that we agree to accept the current version?
Personally, I don't have a strong preference of whether we have to add the fallback, but if I were to choose, I think I'm more convinced by @avikivity's and @haaawk's arguments. If we don't add the additional statement, we will have to rely entirely on the script, and it looks like there were no changes in the columns in the last 2 years (at least there were no changes in the tables in the script), so users didn't need to run it and they might not even be aware that it exists (especially because there is no documentation about it).
All in all, it feels like we're choosing between a minor inconvenience for us and a minor inconvenience for the users, and in such a case, I'm pretty sure we'd prefer the former.

vladzcloudius replied:
@wmitros 
All it takes is to run a script before you restart the service after the upgrade and that will make no extra hacks needed. Everything will continue to work before and after the upgrade.

Our upgrade documentation is per-upgrade version, so the next one that relates to the version with this new column should have the corresponding bullet at the corresponding place. 

> How did you arrive at "a few"? It could be a large number since cluster upgrade is not bounded in time, and there is nothing that will add the missing column automatically.

@avikivity Rephrasing the Mrs. May popular saying: "It's highly unlikely that the number of these requests is going to be greater than 0." 
However in order to cover such an unlikely event one should follow the procedure that we should have in our upgrade notes as I described above.

If the script is executed before the first scylla start after the upgrade then it will be exactly 0 without and fallbacks.

However I see that you have merged it already, so - now we have (more) things to cleanup following  253a764.

haaawk commented on table_helper.cc:
```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
+            _prepared_stmt = std::move(msg_ptr->get_prepared());
+            shared_ptr<cql3::cql_statement> cql_stmt = _prepared_stmt->statement;
+            _insert_stmt = dynamic_pointer_cast<cql3::statements::modification_statement>(cql_stmt);
+            _is_fallback_stmt = true;
```
I have mixed feelings about this state. Is it safe when table_helper::insert (and then table_helper::cache_table_info) is called concurrently multiple times? Will it still behave correctly?

wmitros replied:
Continuations on the same thread (so those who can access the same memory) do not run in parallel. Therefore each time a new value is assigned to _instert_stmt, the _is_fallback_stmt is changed accordingly at the same time.
I should probably add a comment that this is the intended use of the _is_fallback_stmt variable.

haaawk replied:
I guess if we assume that none of the callbacks throws an exception then it should be ok.

wmitros replied:
That's the assumption. Maybe I should have added the noexcept specifier to the lambdas.

haaawk reviewed (approved):
LGTM @avikivity what do you think?

avikivity commented on table_helper.cc:
```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
         shared_ptr<cql3::cql_statement> cql_stmt = _prepared_stmt->statement;
         _insert_stmt = dynamic_pointer_cast<cql3::statements::modification_statement>(cql_stmt);
+        _is_fallback_stmt = false;
+    }).handle_exception([this, &qs] (auto eptr) {
```
We have a rule against unnecessatily-generic lambdas, because they make the lambda harder for IDEs to understand.

avikivity replied:
This can trigger fallbacks for unrelated exceptions, like a bad_alloc. It's unlikely, but better to check for the failure we expect.

avikivity commented on table_helper.hh:
```diff
@@ -95,8 +97,27 @@ public:
+     * @param opt_maker cql_options maker functor
+     * @param opt_maker_fallback cql_options maker functor for the fallback CQL INSERT statement
+     * @param opt_maker_args opt_maker arguments
+     */
```
Please add constraints for all template parameters, both to catch errors, and as a more exact form of documentation.

avikivity replied:
Ugh, you have a constraint. But really, I still can't guess what an opt_maker is.

avikivity commented on table_helper.cc:
```diff
@@ -59,17 +59,31 @@ future<> table_helper::setup_table() const {
+            _prepared_stmt = std::move(msg_ptr->get_prepared());
+            shared_ptr<cql3::cql_statement> cql_stmt = _prepared_stmt->statement;
+            _insert_stmt = dynamic_pointer_cast<cql3::statements::modification_statement>(cql_stmt);
+            _is_fallback_stmt = true;
```
We should age out fallback statements so we notice when the new column is added.

avikivity replied:
Yes, we do. But note we might replace it in the future with more fine-grained invalidation that only invalidates a statement if affected columns are modified.

But it's okay for now.

avikivity commented on tracing/trace_keyspace_helper.hh:
```diff
@@ -142,6 +142,15 @@ private:
+     * @return the relevant cql3::query_options object with the mutation data
+     */
+    static cql3::query_options make_session_mutation_data_fallback(const one_session_records& all_records_handle);
+
```
I think you can use named bind parameters and have just a single query_options thing.

statement: INSERT INTO blah values (blah, blah, username) VALUES (:blah1, :blah2, :username)
fallback:  INSERT INTO blah values (blah, blah) VALUES (:blah1, :blah2)

The :username parameter in the query_options should just be ignored.

wmitros commented:
I've rebased and removed some of the fallback functions in favor of naming the bind parameters and possibly using a subset of them. I've also applied the remaining review comments so the pr should be ready for merge now.

avikivity commented:
Looks good to me. I'll wait a bit for @haaawk and @vladzcloudius to re-review if they want to.

haaawk reviewed (approved):
LGTM
